### DISCUSSION TITLE
Unnecessary conditional, the types have no overlap.

### DISCUSSION THREAD

**Discussion by winstxnhdw** [typescript-eslint/typescript-eslint#7007] [Technical Discussions]
I receive the following lint error `Unnecessary conditional, the types have no overlap.` while trying to type narrow `AccessTokenResponse` to `AccessTokenResponseSuccess`. This is a method the TypeScript team has documented in their `Narrowing` [documentation](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). Is it possible that this is a mistake on the linter?

```ts
type AccessTokenResponse = AccessTokenResponseSuccess | AccessTokenResponseError

interface AccessTokenResponseSuccess {
  access_token: string
  expires_in: number
  refresh_expires_in: number
  refresh_token: string
  token_type: string
  session_state: string
  scope: string
}

interface AccessTokenResponseError {
  error: string
  error_description: string
}

const is_access_token_response = (response: AccessTokenResponse): response is AccessTokenResponseSuccess => {
  return (response as AccessTokenResponseSuccess).access_token !== undefined  // Unnecessary conditional, the types have no overlap.
}

export const generate_access_token = async (): Promise<string> => {
  const access_token_response = await generate_access_token_request()

  if (!is_access_token_response(access_token_response)) {
    throw new Error(access_token_response.error_description)
  }

  return access_token_response.access_token
}
```

**Comment by JoshuaKGoldberg** [comment]
👋 The _Technical Discussions_ category of our GitHub Discussions is meant for _"talking about the architecture and/or APIs of typescript-eslint"_. Please use [GitHub Issues](https://github.com/typescript-eslint/typescript-eslint/issues) if you believe this is a bug in our end, or either [StackOverflow](https://stackoverflow.com) [our new Discord](https://discord.com/invite/FSxKq8Tdyg) if you just have a general question. 

Tip: your code sample doesn't include all the types we'd need to understand your code. When asking for help (in addition to reading contributing guidelines explaining where to direct different types of queries), be sure to always include enough information to fully understand your question.
